crates/chain/src/local_chain.rs

```diff
@@ -17,224 +119,450 @@ impl ChainOracle for LocalChain {
-    pub fn from_blocks<B>(blocks: B) -> Self
-    where
-        B: IntoIterator<Item = BlockId>,
-    {
-        Self {
-            blocks: blocks.into_iter().map(|b| (b.height, b.hash)).collect(),
+    /// Construct a [`LocalChain`] from a given `checkpoint` tip.
+    pub fn from_checkpoint(checkpoint: Arc<CheckPoint>) -> Result<Self, InvalidHeightOrderError> {
+        match Self::default().extend_from_last_agreement(checkpoint) {
+            Ok((chain, _)) => Ok(chain),
+            Err(ExtendError::InvalidHeightOrder(err)) => Err(err),
+            Err(ExtendError::CannotConnect(_)) => {
+                unreachable!("we do not need to connect to an empty chain")
+            }
         }
     }
 
-    /// Get a reference to a map of block height to hash.
-    pub fn blocks(&self) -> &BTreeMap<u32, BlockHash> {
-        &self.blocks
+    /// Constructs a [`LocalChain`] from an iterator of [`BlockId`]s.
+    ///
+    /// The blocks do not need to be in order. If multiple blocks of the same height are introduced,
+    /// the latest block will be used.
+    pub fn from_blocks(blocks: impl IntoIterator<Item = BlockId>) -> Self {
+        let mut chain = Self {
+            checkpoints: blocks
+                .into_iter()
+                .map(|b| {
+                    (
+                        b.height,
+                        Arc::new(CheckPoint {
+                            height: b.height,
+                            hash: b.hash,
+                            prev: None,
+                        }),
+                    )
+                })
+                .collect(),
+        };
+        chain.fix_links(0);
+        chain
     }
 
-    pub fn tip(&self) -> Option<BlockId> {
-        self.blocks
-            .iter()
-            .last()
-            .map(|(&height, &hash)| BlockId { height, hash })
+    /// Get the highest checkpoint.
+    pub fn tip(&self) -> Option<Arc<CheckPoint>> {
+        self.checkpoints.values().last().cloned()
     }
 
-    /// This is like the sparsechain's logic, expect we must guarantee that all invalidated heights
-    /// are to be re-filled.
-    pub fn determine_changeset(&self, update: &Self) -> Result<ChangeSet, UpdateNotConnectedError> {
-        let update = update.as_ref();
-        let update_tip = match update.keys().last().cloned() {
-            Some(tip) => tip,
-            None => return Ok(ChangeSet::default()),
-        };
+    /// Preview changes of updating [`Self`].
+    ///
+    /// To invalidate from a given checkpoint, `update` must contain a checkpoint of the same height
+    /// but different hash.
+    ///
+    /// # Errors
+    ///
+    /// An error will occur if the update does not correctly connect with `self`.
+    ///
+    /// Refer to [module-level documentation] for more.
+    ///
+    /// [module-level documentation]: crate::local_chain
+    pub fn determine_changeset(
+        &self,
+        new_tip: &Arc<CheckPoint>,
+        must_include_height: Option<u32>,
```

Comment:
why isn't the update another `LocalChain`?

Reply:
Formulating a `LocalChain` means we need to think about which checkpoints to include in the update. I think it's a much easier API if the chain-data-source-thread just spits out new checkpoints it is aware of, and the `determine_changeset`. Of course, you can continue to add data to an update `LocalChain` and throw that down a channel, but the internal `BTreeMap` will keep growing.

Reply:
Ok I think that makes sense. so what is the point of `must_include_height`? The doc comments don't mention it.